File: dotnet/VmmTrackerCore/SdpCodec.cs
using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Net;
using System.Text;

namespace VmmTrackerCore;

/// <summary>
/// SDP Compression Codec for WebRTC DataChannel signaling.
///
/// Compresses SDP + ICE candidates into a compact binary format (71-116 bytes)
/// suitable for QR code / base64 text exchange.
///
/// Binary format:
///   [header: 1] [fingerprint: 32] [ufrag_len: 1] [ufrag: U]
///   [pwd_len: 1] [pwd: P] [candidate_count: 1] [candidates: var]
///
/// Each candidate:
///   [flags: 1] [ip: 4|16] [port: 2]
/// </summary>
public static class SdpCodec
{
    private const int ProtocolVersion = 0;

    private static readonly string[] SdpTemplateLines = new[]
    {
        "v=0",
        "o=- 0 0 IN IP4 0.0.0.0",
        "s=-",
        "t=0 0",
        "a=group:BUNDLE 0",
        "m=application 9 UDP/DTLS/SCTP webrtc-datachannel",
        "c=IN IP4 0.0.0.0",
        "a=mid:0",
    };

    /// <summary>
    /// Encode SDP + ICE candidate lines into compact binary.
    /// </summary>
    /// <param name="sdp">Full SDP text</param>
    /// <param name="isOffer">true for offer, false for answer</param>
    /// <param name="iceCandidateLines">Raw ICE candidate attribute lines (a=candidate:...)</param>
    public static byte[] Encode(string sdp, bool isOffer, string[] iceCandidateLines)
    {
        // Extract fingerprint
        var fpMatch = System.Text.RegularExpressions.Regex.Match(sdp, @"a=fingerprint:sha-256 ([0-9A-Fa-f:]+)");
        if (!fpMatch.Success) throw new ArgumentException("No fingerprint found in SDP");
        byte[] fingerprintBytes = HexColonToBytes(fpMatch.Groups[1].Value);

        // Extract ufrag
        var ufragMatch = System.Text.RegularExpressions.Regex.Match(sdp, @"a=ice-ufrag:(.+)");
        if (!ufragMatch.Success) throw new ArgumentException("No ice-ufrag found in SDP");
        string ufrag = ufragMatch.Groups[1].Value.Trim();

        // Extract pwd
        var pwdMatch = System.Text.RegularExpressions.Regex.Match(sdp, @"a=ice-pwd:(.+)");
        if (!pwdMatch.Success) throw new ArgumentException("No ice-pwd found in SDP");
        string pwd = pwdMatch.Groups[1].Value.Trim();

        // Filter host candidates and parse
        var hostCandidates = new List<(bool isIPv6, byte[] ipBytes, ushort port)>();
        foreach (string line in iceCandidateLines)
        {
            if (!line.Contains("typ host")) continue;
            var parsed = ParseCandidateLine(line);
            hostCandidates.Add(parsed);
        }

        // Calculate total size
        byte[] ufragBytes = Encoding.UTF8.GetBytes(ufrag);
        byte[] pwdBytes = Encoding.UTF8.GetBytes(pwd);
        int size = 1 + 32 + 1 + ufragBytes.Length + 1 + pwdBytes.Length + 1;
        foreach (var cand in hostCandidates)
        {
            size += 1 + (cand.isIPv6 ? 16 : 4) + 2;
        }

        byte[] result = new byte[size];
        int offset = 0;

        // Header byte
        byte header = (byte)((ProtocolVersion << 5) | (isOffer ? 0x00 : 0x10));
        result[offset++] = header;

        // Fingerprint (32 bytes)
        Array.Copy(fingerprintBytes, 0, result, offset, 32);
        offset += 32;

        // ufrag (length-prefixed)
        result[offset++] = (byte)ufragBytes.Length;
        Array.Copy(ufragBytes, 0, result, offset, ufragBytes.Length);
        offset += ufragBytes.Length;

        // pwd (length-prefixed)
        result[offset++] = (byte)pwdBytes.Length;
        Array.Copy(pwdBytes, 0, result, offset, pwdBytes.Length);
        offset += pwdBytes.Length;

        // Candidate count
        result[offset++] = (byte)hostCandidates.Count;

        // Candidates
        foreach (var cand in hostCandidates)
        {
            result[offset++] = (byte)(cand.isIPv6 ? 0x01 : 0x00);
            Array.Copy(cand.ipBytes, 0, result, offset, cand.ipBytes.Length);
            offset += cand.ipBytes.Length;
            BinaryPrimitives.WriteUInt16BigEndian(result.AsSpan(offset), cand.port);
            offset += 2;
        }

        return result;
    }

    /// <summary>
    /// Decode compact binary back to a full SDP string.
    /// </summary>
    public static (string sdp, bool isOffer) Decode(byte[] data)
    {
        if (data == null || data.Length < 1)
            throw new ArgumentException("Empty data");

        int offset = 0;

        // Header byte
        byte header = data[offset++];
        int version = (header >> 5) & 0x07;
        if (version != 0)
            throw new ArgumentException($"Unsupported protocol version: {version}");
        bool isOffer = (header & 0x10) == 0;

        // Fingerprint (32 bytes)
        byte[] fpBytes = new byte[32];
        Array.Copy(data, offset, fpBytes, 0, 32);
        offset += 32;
        string fingerprint = BytesToHexColon(fpBytes);

        // ufrag
        int ufragLen = data[offset++];
        string ufrag = Encoding.UTF8.GetString(data, offset, ufragLen);
        offset += ufragLen;

        // pwd
        int pwdLen = data[offset++];
        string pwd = Encoding.UTF8.GetString(data, offset, pwdLen);
        offset += pwdLen;

        // Candidates
        int candCount = data[offset++];
        var candidates = new List<(string ip, ushort port)>();
        for (int i = 0; i < candCount; i++)
        {
            byte flags = data[offset++];
            bool isIPv6 = (flags & 0x01) != 0;
            int ipLen = isIPv6 ? 16 : 4;
            byte[] ipBytes = new byte[ipLen];
            Array.Copy(data, offset, ipBytes, 0, ipLen);
            offset += ipLen;
            ushort port = BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(offset));
            offset += 2;
            string ip = new IPAddress(ipBytes).ToString();
            candidates.Add((ip, port));
        }

        // Build SDP
        string setup = isOffer ? "actpass" : "active";
        var sb = new StringBuilder();

        foreach (string line in SdpTemplateLines)
        {
            sb.Append(line).Append("\r\n");
        }

        sb.Append($"a=ice-ufrag:{ufrag}\r\n");
        sb.Append($"a=ice-pwd:{pwd}\r\n");
        sb.Append($"a=fingerprint:sha-256 {fingerprint}\r\n");
        sb.Append($"a=setup:{setup}\r\n");
        sb.Append("a=sctp-port:5000\r\n");
        sb.Append("a=max-message-size:262144\r\n");

        for (int i = 0; i < candidates.Count; i++)
        {
            var c = candidates[i];
            uint priority = (uint)((126 << 24) | ((65535 - i) << 8) | 255);
            int foundation = i + 1;
            sb.Append($"a=candidate:{foundation} 1 udp {priority} {c.ip} {c.port} typ host generation 0\r\n");
        }

        return (sb.ToString(), isOffer);
    }

    /// <summary>Encode binary data to base64.</summary>
    public static string ToBase64(byte[] data) => Convert.ToBase64String(data);

    /// <summary>Decode base64 string to binary data.</summary>
    public static byte[] FromBase64(string str) => Convert.FromBase64String(str);

    // ── Internal helpers ──

    private static byte[] HexColonToBytes(string hexColon)
    {
        string[] parts = hexColon.Split(':');
        byte[] bytes = new byte[parts.Length];
        for (int i = 0; i < parts.Length; i++)
        {
            bytes[i] = Convert.ToByte(parts[i], 16);
        }
        return bytes;
    }

    private static string BytesToHexColon(byte[] bytes)
    {
        // BitConverter.ToString produces "XX-XX-XX" format, convert to "XX:XX:XX"
        return BitConverter.ToString(bytes).Replace("-", ":");
    }

    private static (bool isIPv6, byte[] ipBytes, ushort port) ParseCandidateLine(string line)
    {
        // Format: "a=candidate:foundation 1 udp priority ip port typ host ..."
        // or: "candidate:foundation 1 udp priority ip port typ host ..."
        string[] parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);

        // Find the start - handle both "a=candidate:..." and "candidate:..."
        int baseIndex = 0;
        for (int i = 0; i < parts.Length; i++)
        {
            if (parts[i].StartsWith("candidate:", StringComparison.Ordinal) ||
                parts[i].StartsWith("a=candidate:", StringComparison.Ordinal))
            {
                baseIndex = i;
                break;
            }
        }

        string ipStr = parts[baseIndex + 4];
        ushort port = ushort.Parse(parts[baseIndex + 5]);
        bool isIPv6 = ipStr.Contains(":");
        byte[] ipBytes = IPAddress.Parse(ipStr).GetAddressBytes();

        return (isIPv6, ipBytes, port);
    }
}

File: web/src/sdp-codec.ts
/**
 * SDP Compression Codec for WebRTC DataChannel signaling.
 *
 * Compresses SDP + ICE candidates into a compact binary format (71-116 bytes)
 * suitable for QR code / base64 text exchange.
 *
 * Binary format:
 *   [header: 1] [fingerprint: 32] [ufrag_len: 1] [ufrag: U]
 *   [pwd_len: 1] [pwd: P] [candidate_count: 1] [candidates: var]
 *
 * Each candidate:
 *   [flags: 1] [ip: 4|16] [port: 2]
 */

const PROTOCOL_VERSION = 0;

const SDP_TEMPLATE_LINES = [
  'v=0',
  'o=- 0 0 IN IP4 0.0.0.0',
  's=-',
  't=0 0',
  'a=group:BUNDLE 0',
  'm=application 9 UDP/DTLS/SCTP webrtc-datachannel',
  'c=IN IP4 0.0.0.0',
  'a=mid:0',
];

/**
 * Encode SDP + ICE candidates into compact binary.
 */
export function encodeSdp(
  sdp: string,
  type: 'offer' | 'answer',
  iceCandidates: RTCIceCandidate[]
): Uint8Array {
  // Extract fingerprint
  const fpMatch = sdp.match(/a=fingerprint:sha-256 ([0-9A-Fa-f:]+)/);
  if (!fpMatch) throw new Error('No fingerprint found in SDP');
  const fingerprintBytes = hexColonToBytes(fpMatch[1]!);

  // Extract ufrag
  const ufragMatch = sdp.match(/a=ice-ufrag:(.+)/);
  if (!ufragMatch) throw new Error('No ice-ufrag found in SDP');
  const ufrag = ufragMatch[1]!.trim();

  // Extract pwd
  const pwdMatch = sdp.match(/a=ice-pwd:(.+)/);
  if (!pwdMatch) throw new Error('No ice-pwd found in SDP');
  const pwd = pwdMatch[1]!.trim();

  // Filter host candidates and parse IP/port
  const hostCandidates = iceCandidates
    .filter((c) => c.candidate && c.candidate.includes('typ host'))
    .map((c) => parseCandidateLine(c.candidate));

  // Calculate total size
  const ufragBytes = new TextEncoder().encode(ufrag);
  const pwdBytes = new TextEncoder().encode(pwd);
  let size = 1 + 32 + 1 + ufragBytes.length + 1 + pwdBytes.length + 1;
  for (const cand of hostCandidates) {
    size += 1 + (cand.isIPv6 ? 16 : 4) + 2;
  }

  const buffer = new ArrayBuffer(size);
  const view = new DataView(buffer);
  const bytes = new Uint8Array(buffer);
  let offset = 0;

  // Header byte
  const header = (PROTOCOL_VERSION << 5) | (type === 'answer' ? 0x10 : 0x00);
  bytes[offset++] = header;

  // Fingerprint (32 bytes)
  bytes.set(fingerprintBytes, offset);
  offset += 32;

  // ufrag (length-prefixed)
  bytes[offset++] = ufragBytes.length;
  bytes.set(ufragBytes, offset);
  offset += ufragBytes.length;

  // pwd (length-prefixed)
  bytes[offset++] = pwdBytes.length;
  bytes.set(pwdBytes, offset);
  offset += pwdBytes.length;

  // Candidate count
  bytes[offset++] = hostCandidates.length;

  // Candidates
  for (const cand of hostCandidates) {
    bytes[offset++] = cand.isIPv6 ? 0x01 : 0x00;
    bytes.set(cand.ipBytes, offset);
    offset += cand.ipBytes.length;
    view.setUint16(offset, cand.port, false); // big-endian
    offset += 2;
  }

  return bytes;
}

/**
 * Decode compact binary back to a full SDP string.
 */
export function decodeSdp(data: Uint8Array): {
  sdp: string;
  type: 'offer' | 'answer';
} {
  if (data.length < 1) throw new Error('Empty data');

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let offset = 0;

  // Header byte
  const header = data[offset++]!;
  const version = (header >> 5) & 0x07;
  if (version !== 0) throw new Error(`Unsupported protocol version: ${version}`);
  const sdpType: 'offer' | 'answer' = (header & 0x10) !== 0 ? 'answer' : 'offer';

  // Fingerprint (32 bytes)
  const fpBytes = data.slice(offset, offset + 32);
  offset += 32;
  const fingerprint = bytesToHexColon(fpBytes);

  // ufrag
  const ufragLen = data[offset++]!;
  const ufrag = new TextDecoder().decode(data.slice(offset, offset + ufragLen));
  offset += ufragLen;

  // pwd
  const pwdLen = data[offset++]!;
  const pwd = new TextDecoder().decode(data.slice(offset, offset + pwdLen));
  offset += pwdLen;

  // Candidates
  const candCount = data[offset++]!;
  const candidates: { ip: string; port: number }[] = [];
  for (let i = 0; i < candCount; i++) {
    const flags = data[offset++]!;
    const isIPv6 = (flags & 0x01) !== 0;
    const ipLen = isIPv6 ? 16 : 4;
    const ipBytes = data.slice(offset, offset + ipLen);
    offset += ipLen;
    const port = view.getUint16(offset, false); // big-endian
    offset += 2;
    const ip = isIPv6 ? ipv6BytesToString(ipBytes) : ipv4BytesToString(ipBytes);
    candidates.push({ ip, port });
  }

  // Build SDP
  const setup = sdpType === 'offer' ? 'actpass' : 'active';
  const lines = [
    ...SDP_TEMPLATE_LINES,
    `a=ice-ufrag:${ufrag}`,
    `a=ice-pwd:${pwd}`,
    `a=fingerprint:sha-256 ${fingerprint}`,
    `a=setup:${setup}`,
    'a=sctp-port:5000',
    'a=max-message-size:262144',
  ];

  for (let i = 0; i < candidates.length; i++) {
    const c = candidates[i]!;
    const priority = (126 << 24) | ((65535 - i) << 8) | 255;
    const foundation = i + 1;
    lines.push(
      `a=candidate:${foundation} 1 udp ${priority >>> 0} ${c.ip} ${c.port} typ host generation 0`
    );
  }

  const sdp = lines.map((l) => l + '\r\n').join('');
  return { sdp, type: sdpType };
}


// ── Internal helpers ──

function hexColonToBytes(hexColon: string): Uint8Array {
  const parts = hexColon.split(':');
  const bytes = new Uint8Array(parts.length);
  for (let i = 0; i < parts.length; i++) {
    bytes[i] = parseInt(parts[i]!, 16);
  }
  return bytes;
}

function bytesToHexColon(bytes: Uint8Array): string {
  const parts: string[] = [];
  for (let i = 0; i < bytes.length; i++) {
    parts.push(bytes[i]!.toString(16).toUpperCase().padStart(2, '0'));
  }
  return parts.join(':');
}

function parseCandidateLine(candidate: string): {
  isIPv6: boolean;
  ipBytes: Uint8Array;
  port: number;
} {
  // candidate format: "candidate:foundation component protocol priority ip port typ host ..."
  // or with "candidate:" prefix already stripped in some APIs
  const parts = candidate.split(/\s+/);
  // Find the index of "candidate:..." or just start from known positions
  // Standard: candidate:foundation 1 udp priority ip port typ host
  let ipIndex: number;
  let portIndex: number;
  if (parts[0]!.startsWith('candidate:')) {
    ipIndex = 4;
    portIndex = 5;
  } else {
    // "candidate" ":" "foundation" "1" "udp" "priority" "ip" "port" ...
    // Shouldn't happen with RTCIceCandidate.candidate but handle gracefully
    ipIndex = 4;
    portIndex = 5;
  }

  const ipStr = parts[ipIndex]!;
  const port = parseInt(parts[portIndex]!, 10);
  const isIPv6 = ipStr.includes(':');
  const ipBytes = isIPv6 ? ipv6StringToBytes(ipStr) : ipv4StringToBytes(ipStr);

  return { isIPv6, ipBytes, port };
}

function ipv4StringToBytes(ip: string): Uint8Array {
  const parts = ip.split('.');
  return new Uint8Array(parts.map((p) => parseInt(p, 10)));
}

function ipv4BytesToString(bytes: Uint8Array): string {
  return Array.from(bytes).join('.');
}

function ipv6StringToBytes(ip: string): Uint8Array {
  // Expand :: notation
  const halves = ip.split('::');
  let groups: string[];
  if (halves.length === 2) {
    const left = halves[0] ? halves[0]!.split(':') : [];
    const right = halves[1] ? halves[1]!.split(':') : [];
    const missing = 8 - left.length - right.length;
    groups = [...left, ...Array(missing).fill('0'), ...right];
  } else {
    groups = ip.split(':');
  }

  const bytes = new Uint8Array(16);
  for (let i = 0; i < 8; i++) {
    const val = parseInt(groups[i]!, 16);
    bytes[i * 2] = (val >> 8) & 0xff;
    bytes[i * 2 + 1] = val & 0xff;
  }
  return bytes;
}

function ipv6BytesToString(bytes: Uint8Array): string {
  const groups: string[] = [];
  for (let i = 0; i < 16; i += 2) {
    const val = (bytes[i]! << 8) | bytes[i + 1]!;
    groups.push(val.toString(16));
  }
  return groups.join(':');
}
